compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1033,6 +1037,11 @@ private boolean isTupleTypeAssignable(BType source, BType target, Set<TypePair>
             unresolvedTypes.add(pair);
         }
 
+        if (target.tag == TypeTags.JSON && source.tag == TypeTags.TUPLE && ((BTupleType) source).tupleTypes.isEmpty()
+        && ((BTupleType) source).restType != null) {
```

Comment:
I don't think this is the correct fix.

The following still doesn't work

```ballerina
    [json, json...] x = [];
    json j = x;
```

Reply:
When calling `isTupleTypeAssignable` for JSON we should probably pass `[json...]` as the target type.

These changes need to be introduced at runtime too.

Reply:
This is not a recursive type problem btw. The following doesn't work either.

```ballerina
    [int, int...] x = [];
    json j = x;
```

Reply:
Tracked by https://github.com/ballerina-platform/ballerina-lang/issues/22244.